src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,355 @@
+    dim_t n[DIM];
+    dim_t inembed[DIM];
+    dim_t onembed[DIM];
+
+    // Full volume dimensions / strides
+    const Point<DIM> zero   = Point<DIM>::ZEROES();
+    const Point<DIM> one    = Point<DIM>::ONES();
+    Point<DIM> fft_size_in  =  in_rect.hi -  in_rect.lo + one;
+    Point<DIM> fft_size_out = out_rect.hi - out_rect.lo + one;
+    size_t num_elements_in  = 1;
+    size_t num_elements_out = 1;
+    for(int i = 0; i < DIM; ++i) {
+      n[i]          = /*(type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] :*/ fft_size_out[i];
+      inembed[i]    = fft_size_in[i];
+      onembed[i]    = fft_size_out[i];
+      num_elements_in  *= fft_size_in[i];
+      num_elements_out *= fft_size_out[i];
+    }
+
+    // Copy input to temporary buffer to perform FFTs one by one
+    DeferredBuffer<INPUT_TYPE, DIM> input_buffer(Rect<DIM>(zero, fft_size_in - one),
+                                                 Memory::GPU_FB_MEM,
+                                                 nullptr /*initial*/,
+                                                 128 /*alignment*/);
+    CHECK_CUDA(cudaMemcpyAsync(input_buffer.ptr(zero), in.ptr(zero), num_elements_in*sizeof(INPUT_TYPE), cudaMemcpyDefault, stream));
+
+    for(auto& ax : axes) {
+      // Create the plan
+      cufftHandle plan;
+      CHECK_CUFFT(cufftCreate(&plan));
+      CHECK_CUFFT(cufftSetAutoAllocation(plan, 0 /*we'll do the allocation*/));
+      CHECK_CUFFT(cufftSetStream(plan, stream));
+
+      // Single axis dimensions / stridfes
+      dim_t size_1d = n[ax];
+      // TODO: batches only correct for DIM <= 3. Fix for N-DIM case
+      dim_t batches = (DIM == 3 && ax == 1) ? n[2] : num_elements_in / n[ax];
+      dim_t istride = 1;
+      dim_t ostride = 1;
+      for(int i = ax+1; i < DIM; ++i) {
+        istride *= fft_size_in[i];
+        ostride *= fft_size_out[i];
+      }
+      dim_t idist = (ax == DIM-1) ? fft_size_in[ax] : 1;
+      dim_t odist = (ax == DIM-1) ? fft_size_out[ax] : 1;
+
+      // Create the plan and allocate a temporary buffer for it if it needs one
+      CHECK_CUFFT(cufftMakePlanMany64(plan, 1, &size_1d, inembed, istride, idist, onembed, ostride, odist, (cufftType)type, batches, &workarea_size));
+
+      DeferredBuffer<uint8_t, 1> workarea_buffer;
+      if(workarea_size > 0) {
+        const Point<1> zero1d(0);
+        workarea_buffer =
+          DeferredBuffer<uint8_t, 1>(Rect<1>(zero1d, Point<1>(workarea_size - 1)),
+                                     Memory::GPU_FB_MEM,
+                                     nullptr /*initial*/,
+                                     128 /*alignment*/);
+        void* workarea = workarea_buffer.ptr(zero1d);
+        CHECK_CUFFT(cufftSetWorkArea(plan, workarea));
+      }
+
+      // TODO: following function only correct for DIM <= 3. Fix for N-DIM case
+      cufft_axes_plan<DIM, INPUT_TYPE, INPUT_TYPE>::execute(plan, input_buffer, input_buffer, out_rect, in_rect, ax, direction);
```

Comment:
I wonder if hiding the details (DIM==3 and DIM<3) in there really makes thing simpler. 
Since the logic differ for DIM==3 and DIM<3 (one or many exec's), I would probably keep everything here, otherwise the logic above (like `dim_t batches = (DIM == 3 && ax == 1) ? n[2] : num_elements_in / n[ax]`) is a little mysterious. Not a big deal.

Reply:
I agree, this is less than ideal. I'd like cuFFT to handle this workflow internally (where we have a 3D input volume but we want to perform 1D FFTs over any axis). For this first implementation, I'd leave it as is with that TODO, but yeah, we should revisit in the future to make it more neat.